src/Network/Wai/SAML2.hs

```diff
@@ -114,9 +116,11 @@ saml2Callback cfg callback app req sendResponse = do
                 -- extract the SAML response from the request body
                 [("SAMLResponse",val)] -> do 
                     result <- validateResponse cfg val
-
+                    let relayState = lookup "RelayState" body
                     -- call the callback
-                    callback result app req sendResponse              
+                    case result of
+                      Left err -> callback (Left err) app req sendResponse
+                      Right assertion -> callback (Right $ AssertionWithState assertion relayState) app req sendResponse
```

Comment:
This is a bit long and verbose, something like this would be better (see in conjunction with other changes):
```suggestion
                    callback (result >>= Result rs) app req sendResponse
```